File: 95-unique-binary-search-trees-ii/95-unique-binary-search-trees-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        
        
        
        def dfs(i,j):
            if i>j:
                return [None]
            arr = []
            for x in range(i,j+1):
                left = dfs(i,x-1);
                right = dfs(x+1,j)
                
                for l in left:
                    for r in right:
                        root = TreeNode(x)
                        root.left , root.right = l,r
                        arr.append(root)
            return arr
        
        return dfs(1,n);

File: 95-unique-binary-search-trees-ii/95-unique-binary-search-trees-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    
    vector<TreeNode*> dfs(int i,int j){
         if(i>j)
            return {nullptr};
            
            vector<TreeNode*> arr;
            
            for(int k=i;k<=j;k++)
            {
                auto left = dfs(i,k-1);
                auto right = dfs(k+1,j);
                
                for(auto &val_l : left)
                {
                    for(auto &val_r : right)
                    {
                        TreeNode *root = new TreeNode(k);
                        root->left = val_l , root->right = val_r;
                        arr.push_back(root);
                    }
                }
            }
            return arr;
    }
    
public:
    vector<TreeNode*> generateTrees(int n) {
        return dfs(1,n);
    }
};